File: ainpc-core-plugin/src/test/java/ro/ainpc/commands/AINPCTabCompleterTest.java
package ro.ainpc.commands;

import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class AINPCTabCompleterTest {

    @Test
    void completesProgressionGuiAliasesWithRelativeFilters() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> subcommandCompletions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"contract", "gui", ""}
        );
        assertTrue(subcommandCompletions.contains("active"));
        assertTrue(subcommandCompletions.contains("tracked"));
        assertTrue(subcommandCompletions.contains("archived"));
        assertFalse(subcommandCompletions.contains("quest"));

        List<String> directCompletions = completer.onTabComplete(
            null,
            new TestCommand("contract"),
            "contract",
            new String[] {"gui", ""}
        );
        assertTrue(directCompletions.contains("current"));
        assertTrue(directCompletions.contains("failed"));
        assertFalse(directCompletions.contains("ritual"));
    }

    @Test
    void completesGlobalProgressionGuiWithKindFilters() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> completions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"gui", "progresii", ""}
        );
        assertTrue(completions.contains("contract"));
        assertTrue(completions.contains("ritual"));
        assertTrue(completions.contains("active"));

        List<String> directProgressionCompletions = completer.onTabComplete(
            null,
            new TestCommand("progression"),
            "progression",
            new String[] {"gui", ""}
        );
        assertTrue(directProgressionCompletions.contains("contract"));
        assertTrue(directProgressionCompletions.contains("ritual"));
        assertTrue(directProgressionCompletions.contains("active"));
    }

    @Test
    void completesQuestAnchorsWithTemplateOrCodeHint() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> completions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"quest", "anchors", "all", ""}
        );

        assertTrue(completions.contains("<templateId|questCode>"));
    }

    @Test
    void completesAuditWandMode() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> completions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"audit", ""}
        );

        assertTrue(completions.contains("wand"));
    }

    @Test
    void completesAuditQuestStrictOption() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> completions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"audit", "quest", ""}
        );

        assertTrue(completions.contains("strict"));
        assertTrue(completions.contains("full"));
    }

    @Test
    void completesPatchPlannerActionsAndFallbackRegionHint() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> actions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"patch", ""}
        );
        assertTrue(actions.contains("analyze"));
        assertTrue(actions.contains("plan"));
        assertTrue(actions.contains("validate"));

        List<String> regions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"patch", "plan", ""}
        );
        assertTrue(regions.contains("<regionId>"));

        List<String> professions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"patch", "plan", "demo_sat", "8", ""}
        );
        assertTrue(professions.contains("blacksmith"));
        assertTrue(professions.contains("blacksmith,farmer"));
    }

    @Test
    void completesNpcBindingRepairTargets() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> targets = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"repair", ""}
        );
        assertTrue(targets.contains("npc-bindings"));
        assertTrue(targets.contains("mapping-metadata"));

        List<String> modes = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"repair", "npc-bindings", ""}
        );
        assertTrue(modes.contains("dryrun"));
        assertTrue(modes.contains("apply"));
    }

    @Test
    void completesMappingWandAndDraftCommands() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> wandActions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"wand", ""}
        );
        assertTrue(wandActions.contains("inspect"));
        assertTrue(wandActions.contains("reset"));

        List<String> modes = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"wand", "mode", ""}
        );
        assertTrue(modes.contains("region"));
        assertTrue(modes.contains("node"));
        assertTrue(modes.contains("quest_anchor"));

        List<String> resetTargets = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"wand", "reset", ""}
        );
        assertTrue(resetTargets.contains("pos1"));
        assertTrue(resetTargets.contains("pos2"));
        assertTrue(resetTargets.contains("point"));
        assertTrue(resetTargets.contains("all"));

        List<String> mapActions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"map", ""}
        );
        assertTrue(mapActions.contains("region"));
        assertTrue(mapActions.contains("place"));
        assertTrue(mapActions.contains("npc_bind"));
        assertTrue(mapActions.contains("quest_anchor"));
        assertTrue(mapActions.contains("preview"));
        assertTrue(mapActions.contains("confirm"));
    }

    @Test
    void completesQuestAnchorDraftShape() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> selectors = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"map", "quest_anchor", ""}
        );
        assertTrue(selectors.contains("tracked"));
        assertTrue(selectors.contains("current"));
        assertTrue(selectors.contains("player:self"));

        List<String> objectiveHint = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"map", "quest_anchor", "tracked", ""}
        );
        assertTrue(objectiveHint.contains("<objective_id>"));

        List<String> objectiveTypes = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"map", "quest_anchor", "tracked", "objective_1", ""}
        );
        assertTrue(objectiveTypes.contains("visit_place"));
        assertTrue(objectiveTypes.contains("inspect_node"));
    }

    @Test
    void completesHouseholdInspectionActions() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> completions = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"world", "household", ""}
        );

        assertTrue(completions.contains("status"));
        assertTrue(completions.contains("place"));
        assertTrue(completions.contains("resident"));
        assertTrue(completions.contains("list"));
    }

    @Test
    void completesMigrationHouseholdActions() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> targets = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"migration", ""}
        );
        assertTrue(targets.contains("households"));

        List<String> modes = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"migration", "households", ""}
        );
        assertTrue(modes.contains("dryrun"));
        assertTrue(modes.contains("apply"));
    }

    @Test
    void completesHouseholdRepairActions() {
        AINPCTabCompleter completer = new AINPCTabCompleter(null);

        List<String> targets = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"repair", ""}
        );
        assertTrue(targets.contains("households"));
        assertTrue(targets.contains("batch"));

        List<String> modes = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"repair", "households", ""}
        );
        assertTrue(modes.contains("dryrun"));
        assertTrue(modes.contains("apply"));

        List<String> batchModes = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"repair", "batch", "settlement:sat:abc", ""}
        );
        assertTrue(batchModes.contains("dryrun"));
        assertTrue(batchModes.contains("apply"));
        assertTrue(batchModes.contains("inspect"));
        assertTrue(batchModes.contains("mark-steps"));
        assertTrue(batchModes.contains("mark-failed"));

        List<String> batchKeys = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"repair", "batch", ""}
        );
        assertTrue(batchKeys.contains("list"));

        List<String> batchListFilters = completer.onTabComplete(
            null,
            new TestCommand("ainpc"),
            "ainpc",
            new String[] {"repair", "batch", "list", ""}
        );
        assertTrue(batchListFilters.contains("problem"));
        assertTrue(batchListFilters.contains("failed"));
        assertTrue(batchListFilters.contains("rolled_back"));
    }

    private static final class TestCommand extends Command {
        private TestCommand(String name) {
            super(name);
        }

        @Override
        public boolean execute(CommandSender sender, String commandLabel, String[] args) {
            return true;
        }
    }
}

File: ainpc-core-plugin/src/test/kotlin/ro/ainpc/commands/AINPCTabCompleterTest.kt
package ro.ainpc.commands

import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.lang.reflect.Proxy

class AINPCTabCompleterTest {
    private val sender: CommandSender = Proxy.newProxyInstance(
        CommandSender::class.java.classLoader,
        arrayOf(CommandSender::class.java)
    ) { _, method, _ ->
        when (method.returnType) {
            java.lang.Boolean.TYPE -> false
            java.lang.Integer.TYPE -> 0
            java.lang.Long.TYPE -> 0L
            java.lang.Double.TYPE -> 0.0
            java.lang.Float.TYPE -> 0f
            java.lang.Short.TYPE -> 0.toShort()
            java.lang.Byte.TYPE -> 0.toByte()
            java.lang.Character.TYPE -> 0.toChar()
            String::class.java -> "test"
            else -> null
        }
    } as CommandSender

    private fun tab(completer: AINPCTabCompleter, command: String, vararg args: String): List<String> {
        return completer.onTabComplete(sender, TestCommand(command), command, args.toList().toTypedArray())
    }

    @Test
    fun completesProgressionGuiAliasesWithRelativeFilters() {
        val completer = AINPCTabCompleter(null)
        val sub = tab(completer, "ainpc", "contract", "gui", "")
        assertTrue(sub.contains("active"))
        assertTrue(sub.contains("tracked"))
        assertTrue(sub.contains("archived"))
        assertFalse(sub.contains("quest"))

        val direct = tab(completer, "contract", "gui", "")
        assertTrue(direct.contains("current"))
        assertTrue(direct.contains("failed"))
        assertFalse(direct.contains("ritual"))
    }

    @Test
    fun completesGlobalProgressionGuiWithKindFilters() {
        val completer = AINPCTabCompleter(null)
        val c1 = tab(completer, "ainpc", "gui", "progresii", "")
        assertTrue(c1.contains("contract"))
        assertTrue(c1.contains("ritual"))
        assertTrue(c1.contains("active"))

        val c2 = tab(completer, "progression", "gui", "")
        assertTrue(c2.contains("contract"))
        assertTrue(c2.contains("ritual"))
        assertTrue(c2.contains("active"))
    }

    @Test
    fun completesQuestAnchorsWithTemplateOrCodeHint() {
        val completer = AINPCTabCompleter(null)
        assertTrue(tab(completer, "ainpc", "quest", "anchors", "all", "").contains("<templateId|questCode>"))
    }

    @Test
    fun completesAuditWandMode() {
        val completer = AINPCTabCompleter(null)
        assertTrue(tab(completer, "ainpc", "audit", "").contains("wand"))
    }

    @Test
    fun completesAuditQuestStrictOption() {
        val completer = AINPCTabCompleter(null)
        val c = tab(completer, "ainpc", "audit", "quest", "")
        assertTrue(c.contains("strict"))
        assertTrue(c.contains("full"))
    }

    @Test
    fun completesNpcDemoInspectionAndRepairCommands() {
        val completer = AINPCTabCompleter(null)

        assertTrue(tab(completer, "ainpc", "").contains("list"))
        assertTrue(tab(completer, "ainpc", "").contains("info"))
        assertTrue(tab(completer, "ainpc", "").contains("duplicates"))
        assertTrue(tab(completer, "ainpc", "info", "").contains("nearest"))
        assertTrue(tab(completer, "ainpc", "audit", "").contains("npc"))
        assertTrue(tab(completer, "ainpc", "repair", "").contains("duplicates"))
        assertTrue(tab(completer, "ainpc", "repair", "duplicates", "").contains("dryrun"))
        assertTrue(tab(completer, "ainpc", "debugdump", "").contains("npc"))
    }

    @Test
    fun completesRoutineStatusNearestForDemoFlow() {
        val completer = AINPCTabCompleter(null)

        assertTrue(tab(completer, "ainpc", "routine", "").contains("status"))
        assertTrue(tab(completer, "ainpc", "routine", "").contains("tick"))
        assertTrue(tab(completer, "ainpc", "routine", "status", "").contains("nearest"))
    }

    @Test
    fun completesPatchPlannerActionsAndFallbackRegionHint() {
        val completer = AINPCTabCompleter(null)
        val a = tab(completer, "ainpc", "patch", "")
        assertTrue(a.contains("analyze"))
        assertTrue(a.contains("plan"))
        assertTrue(a.contains("validate"))
        assertTrue(tab(completer, "ainpc", "patch", "plan", "").contains("<regionId>"))
        val p = tab(completer, "ainpc", "patch", "plan", "demo_sat", "8", "")
        assertTrue(p.contains("blacksmith"))
        assertTrue(p.contains("blacksmith,farmer"))
    }

    @Test
    fun completesNpcBindingRepairTargets() {
        val completer = AINPCTabCompleter(null)
        val t = tab(completer, "ainpc", "repair", "")
        assertTrue(t.contains("npc-bindings"))
        assertTrue(t.contains("mapping-metadata"))
        val m = tab(completer, "ainpc", "repair", "npc-bindings", "")
        assertTrue(m.contains("dryrun"))
        assertTrue(m.contains("apply"))
    }

    @Test
    fun completesMappingWandAndDraftCommands() {
        val completer = AINPCTabCompleter(null)
        assertTrue(tab(completer, "ainpc", "wand", "").contains("inspect"))
        assertTrue(tab(completer, "ainpc", "wand", "").contains("reset"))
        val modes = tab(completer, "ainpc", "wand", "mode", "")
        assertTrue(modes.contains("region"))
        assertTrue(modes.contains("node"))
        assertTrue(modes.contains("quest_anchor"))
        val reset = tab(completer, "ainpc", "wand", "reset", "")
        assertTrue(reset.contains("pos1"))
        assertTrue(reset.contains("pos2"))
        assertTrue(reset.contains("point"))
        assertTrue(reset.contains("all"))
        val map = tab(completer, "ainpc", "map", "")
        assertTrue(map.contains("region"))
        assertTrue(map.contains("place"))
        assertTrue(map.contains("npc_bind"))
        assertTrue(map.contains("quest_anchor"))
        assertTrue(map.contains("preview"))
        assertTrue(map.contains("confirm"))
    }

    @Test
    fun completesQuestAnchorDraftShape() {
        val completer = AINPCTabCompleter(null)
        val selectors = tab(completer, "ainpc", "map", "quest_anchor", "")
        assertTrue(selectors.contains("tracked"))
        assertTrue(selectors.contains("current"))
        assertTrue(selectors.contains("player:self"))
        assertTrue(tab(completer, "ainpc", "map", "quest_anchor", "tracked", "").contains("<objective_id>"))
        val types = tab(completer, "ainpc", "map", "quest_anchor", "tracked", "objective_1", "")
        assertTrue(types.contains("visit_place"))
        assertTrue(types.contains("inspect_node"))
    }

    @Test
    fun completesHouseholdInspectionActions() {
        val completer = AINPCTabCompleter(null)
        val c = tab(completer, "ainpc", "world", "household", "")
        assertTrue(c.contains("status"))
        assertTrue(c.contains("place"))
        assertTrue(c.contains("resident"))
        assertTrue(c.contains("list"))
    }

    @Test
    fun completesDemoReadinessCommand() {
        val completer = AINPCTabCompleter(null)
        assertTrue(tab(completer, "ainpc", "").contains("demo"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("status"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("check"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("readiness"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("next"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("blockers"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("todo"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("definition"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("criteria"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("meaning"))
        assertFalse(tab(completer, "ainpc", "demo", "definition", "").contains("<regionId>"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("script"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("guide"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("flow"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("phases"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("checklist"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("evidence"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("runbook"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("smoke"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("summary"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("commands"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("cmds"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("copy"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("restart"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("reloadcheck"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("persistence"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("experimental"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("exp"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("maxpack"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("experimental5"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("exp5"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("fivepack"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("experimental25"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("exp25"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("task25"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("experimental25deep"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("exp25deep"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("deep25"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("experimental25ops"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("exp25ops"))
        assertTrue(tab(completer, "ainpc", "demo", "").contains("ops25"))
        assertTrue(tab(completer, "ainpc", "demo", "script", "").contains("<regionId>"))
        assertTrue(tab(completer, "ainpc", "demo", "script", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "phases", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "evidence", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "runbook", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "smoke", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "summary", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "commands", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "experimental", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "experimental5", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "experimental25", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "experimental25deep", "demo_sat", "").contains("<player>"))
        assertTrue(tab(completer, "ainpc", "demo", "experimental25ops", "demo_sat", "").contains("<player>"))
        assertFalse(tab(completer, "ainpc", "demo", "restart", "demo_sat", "").contains("<player>"))
    }

    @Test
    fun completesMigrationHouseholdActions() {
        val completer = AINPCTabCompleter(null)
        assertTrue(tab(completer, "ainpc", "migration", "").contains("households"))
        val modes = tab(completer, "ainpc", "migration", "households", "")
        assertTrue(modes.contains("dryrun"))
        assertTrue(modes.contains("apply"))
    }

    @Test
    fun completesHouseholdRepairActions() {
        val completer = AINPCTabCompleter(null)
        val t = tab(completer, "ainpc", "repair", "")
        assertTrue(t.contains("households"))
        assertTrue(t.contains("batch"))
        val modes = tab(completer, "ainpc", "repair", "households", "")
        assertTrue(modes.contains("dryrun"))
        assertTrue(modes.contains("apply"))
        val batchModes = tab(completer, "ainpc", "repair", "batch", "settlement:sat:abc", "")
        assertTrue(batchModes.contains("dryrun"))
        assertTrue(batchModes.contains("apply"))
        assertTrue(batchModes.contains("inspect"))
        assertTrue(batchModes.contains("mark-steps"))
        assertTrue(batchModes.contains("mark-failed"))
        assertTrue(tab(completer, "ainpc", "repair", "batch", "").contains("list"))
        val filters = tab(completer, "ainpc", "repair", "batch", "list", "")
        assertTrue(filters.contains("problem"))
        assertTrue(filters.contains("failed"))
        assertTrue(filters.contains("rolled_back"))
    }

    private class TestCommand(name: String) : Command(name) {
        override fun execute(sender: CommandSender, commandLabel: String, args: Array<out String>): Boolean = true
    }
}
